internal/dataplane/parser/translate_udproute.go

```diff
@@ -0,0 +1,187 @@
+	// first we grab the spec and gather some metdata about the object
+	spec := udproute.Spec
+
+	// validation for UDPRoutes will happen at a higher layer, but in spite of that we run
+	// validation at this level as well as a fallback so that if routes are posted which
+	// are invalid somehow make it past validation (e.g. the webhook is not enabled) we can
+	// at least try to provide a helpful message about the situation in the manager logs.
+	if len(spec.Rules) == 0 {
+		return fmt.Errorf("no rules provided")
+	}
+
+	// each rule may represent a different set of backend services that will be accepting
+	// traffic, so we make separate routes and Kong services for every present rule.
+	for ruleNumber, rule := range spec.Rules {
+		// TODO: add this to a generic UDPRoute validation, and then we should probably
+		//       simply be calling validation on each udproute object at the begininning
+		//       of the topmost list.
+		if len(rule.BackendRefs) == 0 {
+			return fmt.Errorf("missing backendRef in rule")
+		}
+
+		// TODO: support multiple backend refs
+		if len(rule.BackendRefs) > 1 {
+			return fmt.Errorf("multiple backendRefs are not yet supported")
+		}
+
+		// determine the routes needed to route traffic to services for this rule
+		routes, err := generateKongRoutesFromUDPRouteRule(udproute, ruleNumber, rule)
+		if err != nil {
+			return err
+		}
+
+		// create a service and attach the routes to it
+		service := generateKongServiceFromUDPRouteBackendRef(result, udproute, rule.BackendRefs[0])
+		service.Routes = append(service.Routes, routes...)
+
+		// cache the service to avoid duplicates in further loop iterations
+		result.ServiceNameToServices[*service.Service.Name] = service
+	}
+
+	return nil
+}
+
+// -----------------------------------------------------------------------------
+// Translate UDPRoute - Utils
+// -----------------------------------------------------------------------------
+
+// generateKongRoutesFromUDPRouteRule converts an UDPRoute rule to one or more
+// Kong Route objects to route traffic to services.
+func generateKongRoutesFromUDPRouteRule(udproute *gatewayv1alpha2.UDPRoute, ruleNumber int,
+	rule gatewayv1alpha2.UDPRouteRule) ([]kongstate.Route, error) {
+	// gather the k8s object information and hostnames from the udproute
+	objectInfo := util.FromK8sObject(udproute)
+
+	var routes []kongstate.Route
+	if len(rule.Matches) > 0 {
+		// As of 2022-03-04, matches are supported only in experimental CRDs. if you apply a UDPRoute with matches against
+		// the stable CRDs, the matches disappear into the ether (only if doing it via client-go, kubectl rejects them)
+		// We do not intend to implement these until they are stable per https://github.com/Kong/kubernetes-ingress-controller/issues/2087#issuecomment-1079053290
+		return routes, fmt.Errorf("UDPRoute Matches are not yet supported")
+	}
+	if len(rule.BackendRefs) > 1 {
```

Comment:
I realized this when working on `HTTPRoute` recently and added it there: we should have an error that can be thrown for when there's _no_ backendRefs provided also.

Reply:
ebcca76b